src/lib/operators/join_index.cpp

```diff
@@ -319,6 +367,59 @@ void JoinIndex::_append_matches(const BaseIndex::Iterator& range_begin, const Ba
   }
 }
 
+void JoinIndex::_append_matches_dereferenced(const ChunkID& probe_chunk_id, const ChunkOffset& probe_chunk_offset,
+                                             const PosList& index_table_matches) {
+  for (const auto& index_side_row_id : index_table_matches) {
+    _probe_pos_list->emplace_back(RowID{probe_chunk_id, probe_chunk_offset});
+    _index_pos_list->emplace_back((index_side_row_id));
+    _index_pos_dereferenced.emplace_back(true);
+  }
+}
+
+void JoinIndex::_append_matches_non_inner(const bool is_semi_or_anti_join) {
+  // For Full Outer and Left Join we need to add all unmatched rows for the probe side
+  if (_mode == JoinMode::Left || _mode == JoinMode::FullOuter) {
+    for (ChunkID probe_chunk_id = ChunkID{0}; probe_chunk_id < _probe_input_table->chunk_count(); ++probe_chunk_id) {
+      for (ChunkOffset chunk_offset{0}; chunk_offset < _probe_matches[probe_chunk_id].size(); ++chunk_offset) {
+        if (!_probe_matches[probe_chunk_id][chunk_offset]) {
+          _probe_pos_list->emplace_back(RowID{probe_chunk_id, chunk_offset});
+          _index_pos_list->emplace_back(NULL_ROW_ID);
```

Comment:
Aren't you violating the `_probe_pos_list->size()` == `_index_pos_list->size()` == `_index_pos_dereferenced.size()` invariant here?